clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java

```diff
@@ -3505,6 +3510,61 @@ protected Set<String> maybeSendIngestionHeartbeat() {
     return failedPartitions;
   }
 
+  /**
+   *  Resubscribe operation by passing new version role and partition role to {@link AggKafkaConsumerService}. The action
+   *  for leader and follower replica will be handled differently.
+   */
+  @Override
+  protected void resubscribe(PartitionConsumptionState partitionConsumptionState) throws InterruptedException {
+    if (isLeader(partitionConsumptionState)) {
+      resubscribeAsLeader(partitionConsumptionState);
+    } else {
+      resubscribeAsFollower(partitionConsumptionState);
+    }
+  }
+
+  protected void resubscribeAsFollower(PartitionConsumptionState partitionConsumptionState)
+      throws InterruptedException {
+    int partition = partitionConsumptionState.getPartition();
+    consumerUnSubscribe(versionTopic, partitionConsumptionState);
+    waitForAllMessageToBeProcessedFromTopicPartition(
+        new PubSubTopicPartitionImpl(versionTopic, partition),
+        partitionConsumptionState);
+    LOGGER.info(
+        "Follower replica: {} unsubscribe finished for future resubscribe.",
+        partitionConsumptionState.getReplicaId());
+    PubSubTopicPartition followerTopicPartition = new PubSubTopicPartitionImpl(versionTopic, partition);
+    long latestProcessedLocalVersionTopicOffset = partitionConsumptionState.getLatestProcessedLocalVersionTopicOffset();
+    consumerSubscribe(followerTopicPartition, false, latestProcessedLocalVersionTopicOffset, localKafkaServer);
+    LOGGER.info(
+        "Follower replica: {} resubscribe to offset: {}",
+        partitionConsumptionState.getReplicaId(),
+        latestProcessedLocalVersionTopicOffset);
+  }
+
+  protected void resubscribeAsLeader(PartitionConsumptionState partitionConsumptionState) throws InterruptedException {
+    OffsetRecord offsetRecord = partitionConsumptionState.getOffsetRecord();
+    PubSubTopic leaderTopic = offsetRecord.getLeaderTopic(pubSubTopicRepository);
+    int partition = partitionConsumptionState.getPartition();
+    consumerUnSubscribe(leaderTopic, partitionConsumptionState);
```

Comment:
Just curious, if we don't pass in the previous version role, how do we know which pool to unsubscribe from?

Reply:
We do not need this information to unsubscribe. For each host, there will only one unique <version_topic, topic_partition> to be subscribed at any time, no matter which consumer pool it belongs to.